File: src/tables/cbdt.cpp
#include "src/algo.h"
#include "src/parser.h"
#include "tables.h"

void parseCbdt(const QVector<CblcIndex> &cblcLocations, Parser &parser)
{
    const auto start = parser.offset();

    const auto majorVersion = parser.read<UInt16>("Major version");
    const auto minorVersion = parser.read<UInt16>("Minor version");

    if (!((majorVersion == 2 || majorVersion == 3) && minorVersion == 0)) {
        throw "invalid table version";
    }

    for (const auto &loca : cblcLocations) {
        parser.jumpTo(start + loca.range.start);
        parser.beginGroup(QString("Bitmap Format %1").arg(loca.imageFormat));

        if (loca.imageFormat == 1) {
            parseSbitSmallGlyphMetrics(parser);
            parser.readBytes(loca.range.size() - 5, "Byte-aligned bitmap data");
        } else if (loca.imageFormat == 2) {
            parseSbitSmallGlyphMetrics(parser);
            parser.readBytes(loca.range.size() - 5, "Bit-aligned bitmap data");
        } else if (loca.imageFormat == 5) {
            parser.readBytes(loca.range.size(), "Bit-aligned bitmap data");
        } else if (loca.imageFormat == 6) {
            parseSbitBigGlyphMetrics(parser);
            parser.readBytes(loca.range.size() - 5, "Byte-aligned bitmap data");
        } else if (loca.imageFormat == 7) {
            parseSbitBigGlyphMetrics(parser);
            parser.readBytes(loca.range.size() - 5, "Bit-aligned bitmap data");
        } else if (loca.imageFormat == 8) {
            parseSbitSmallGlyphMetrics(parser);
            parser.read<UInt8>("Pad");
            const auto count = parser.read<UInt16>("Number of components");
            for (int i = 0; i < count; ++i) {
                parser.beginGroup("Ebdt component");
                parser.read<GlyphId>("Glyph ID");
                parser.read<Int8>("X-axis offset");
                parser.read<Int8>("Y-axis offset");
                parser.endGroup();
            }
        } else if (loca.imageFormat == 9) {
            parseSbitBigGlyphMetrics(parser);
            const auto count = parser.read<UInt16>("Number of components");
            for (int i = 0; i < count; ++i) {
                parser.beginGroup("Ebdt component");
                parser.read<GlyphId>("Glyph ID");
                parser.read<Int8>("X-axis offset");
                parser.read<Int8>("Y-axis offset");
                parser.endGroup();
            }
        } else if (loca.imageFormat == 17) {
            parseSbitSmallGlyphMetrics(parser);
            const auto len = parser.read<UInt32>("Length of data");
            parser.readBytes(len, "Raw PNG data");
        } else if (loca.imageFormat == 18) {
            parseSbitBigGlyphMetrics(parser);
            const auto len = parser.read<UInt32>("Length of data");
            parser.readBytes(len, "Raw PNG data");
        } else if (loca.imageFormat == 19) {
            const auto len = parser.read<UInt32>("Length of data");
            parser.readBytes(len, "Raw PNG data");
        }

        parser.endGroup();
    }
}

File: core/src/tables/cbdt.rs
use crate::parser::*;
use crate::{Error, Result};
use super::cblc;

pub fn parse(indices: &[super::cblc::CblcIndex], parser: &mut Parser) -> Result<()> {
    let start = parser.offset();

    let major_version = parser.read::<u16>("Major version")?;
    let minor_version = parser.read::<u16>("Minor version")?;
    if !((major_version == 2 || major_version == 3) && minor_version == 0) {
        return Err(Error::InvalidTableVersion);
    }

    for loca in indices {
        parser.jump_to(start + loca.range.start)?;
        parser.begin_group(format!("Bitmap Format {}", loca.image_format));

        match loca.image_format {
            1 => {
                cblc::parse_small_glyph_metrics(parser)?;
                parser.read_bytes(loca.range.len() - 5, "Byte-aligned bitmap data")?;
            }
            2 => {
                cblc::parse_small_glyph_metrics(parser)?;
                parser.read_bytes(loca.range.len() - 5, "Bit-aligned bitmap data")?;
            }
            5 => {
                parser.read_bytes(loca.range.len(), "Bit-aligned bitmap data")?;
            }
            6 => {
                cblc::parse_big_glyph_metrics(parser)?;
                parser.read_bytes(loca.range.len() - 5, "Byte-aligned bitmap data")?;
            }
            7 => {
                cblc::parse_big_glyph_metrics(parser)?;
                parser.read_bytes(loca.range.len() - 5, "Bit-aligned bitmap data")?;
            }
            8 => {
                cblc::parse_small_glyph_metrics(parser)?;
                parser.read::<u8>("Pad")?;
                let count = parser.read::<u16>("Number of components")?;
                for _ in 0..count {
                    parser.begin_group("Ebdt component");
                    parser.read::<GlyphId>("Glyph ID")?;
                    parser.read::<i8>("X-axis offset")?;
                    parser.read::<i8>("Y-axis offset")?;
                    parser.end_group();
                }
            }
            9 => {
                cblc::parse_big_glyph_metrics(parser)?;
                let count = parser.read::<u16>("Number of components")?;
                for _ in 0..count {
                    parser.begin_group("Ebdt component");
                    parser.read::<GlyphId>("Glyph ID")?;
                    parser.read::<i8>("X-axis offset")?;
                    parser.read::<i8>("Y-axis offset")?;
                    parser.end_group();
                }
            }
            17 => {
                cblc::parse_small_glyph_metrics(parser)?;
                let len = parser.read::<u32>("Length of data")? as usize;
                parser.read_bytes(len, "Raw PNG data")?;
            }
            18 => {
                cblc::parse_big_glyph_metrics(parser)?;
                let len = parser.read::<u32>("Length of data")? as usize;
                parser.read_bytes(len, "Raw PNG data")?;
            }
            19 => {
                let len = parser.read::<u32>("Length of data")? as usize;
                parser.read_bytes(len, "Raw PNG data")?;
            }
            _ => {}
        }

        parser.end_group();
    }

    Ok(())
}
